anitya/db/events.py

```diff
@@ -19,50 +19,68 @@
 from sqlalchemy import event
 
 from anitya.lib import plugins
-from .meta import Session
 from .models import Project
 
 
 _log = logging.getLogger(__name__)
 
 
-@event.listens_for(Session, "before_flush")
-def set_ecosystem(session, flush_context, instances):
+def _set_ecosystem(project, backend, homepage):
     """
-    An SQLAlchemy event listener that sets the ecosystem for a project if it's null.
+    Set ecosystem to correct value. Priority is as follows:
+        1. Ecosystem is set to backend if backend is associated with ecosystem
+        2. Ecosystem is set to homepage
 
     Args:
-        session (sqlalchemy.orm.session.Session): The session that is about to be committed.
-        flush_context (sqlalchemy.orm.session.UOWTransaction): Unused.
-        instances (object): deprecated and unused
+        project (models.Project): Instance of project
+        backend (str): value of backend
+        homepage (str): value of homepage
+    """
+    ecosystems = [
+        e
+        for e in plugins.ECOSYSTEM_PLUGINS.get_plugins()
+        if e.default_backend == backend
+    ]
+    if ecosystems:
+        project.ecosystem_name = ecosystems[0].name
+    else:
+        project.ecosystem_name = homepage
+    _log.info(
+        "Settings the ecosystem on %r to %s", project.name, project.ecosystem_name
+    )
+
+
+@event.listens_for(Project.backend, "set", raw=True)
+def set_ecosystem_backend(target, value, old, initiator):
+    """
+    An SQLAlchemy event listener that sets the ecosystem for a project if backend is changed.
 
-    Raises:
-        ValueError: If the ecosystem_name isn't valid.
+    Args:
+        target (sqlalchemy.orm.state.InstanceStace): Instance of the object where
+                change is happening.
+        value (str): The new value of backend.
+        old (str): The old value of backend.
+        initiator (sqlalchemy.orm.attributes.Event): The event object that is initiating this
+                transition.
+    """
+    if not value == old:
```

Comment:
Why not `if value != old:`? :wink:

Reply:
I'm probably just confused from java and I thought this is not working in Python. :-D